File: cli/render.test.ts
import { readJsonRpcSse } from '../src/a2a/client';
import type { StreamEvent } from '../src/a2a/types';
import type { AgentCard } from '../src/a2a/agent-card';
import { renderCard, renderChat, renderTask, type Io } from './render';
import { describe, expect, it } from 'bun:test';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

const FIXTURE = readFileSync(join(import.meta.dir, 'fixtures/chat-stream.sse.txt'), 'utf8');

function captureIo(): { io: Io; out: () => string; err: () => string } {
  const outChunks: string[] = [];
  const errChunks: string[] = [];
  return {
    io: { out: (t) => outChunks.push(t), err: (t) => errChunks.push(t) },
    out: () => outChunks.join(''),
    err: () => errChunks.join(''),
  };
}

/** A ReadableStream of the string, chunked to exercise frame reassembly. */
function streamFromString(s: string, chunkSize: number): ReadableStream<Uint8Array> {
  const bytes = new TextEncoder().encode(s);
  let offset = 0;
  return new ReadableStream({
    pull(controller) {
      if (offset >= bytes.length) {
        controller.close();
        return;
      }
      controller.enqueue(bytes.subarray(offset, offset + chunkSize));
      offset += chunkSize;
    },
  });
}

async function collect<T>(gen: AsyncIterable<T>): Promise<T[]> {
  const out: T[] = [];
  for await (const x of gen) out.push(x);
  return out;
}

describe('readJsonRpcSse (recorded transcript)', () => {
  it('parses every frame from the recorded stream', async () => {
    const events = await collect(readJsonRpcSse(streamFromString(FIXTURE, 4096)));
    const kinds = events.map((e) => e.kind);
    expect(kinds[0]).toBe('task');
    expect(kinds).toContain('status-update');
    expect(kinds).toContain('artifact-update');
    const terminal = events.find(
      (e): e is Extract<StreamEvent, { kind: 'status-update' }> =>
        e.kind === 'status-update' && e.status.state === 'completed'
    );
    expect(terminal?.final).toBe(true);
  });

  it('reassembles frames split across tiny chunk boundaries', async () => {
    const whole = await collect(readJsonRpcSse(streamFromString(FIXTURE, 4096)));
    const split = await collect(readJsonRpcSse(streamFromString(FIXTURE, 7)));
    expect(split.length).toBe(whole.length);
    expect(split.map((e) => e.kind)).toEqual(whole.map((e) => e.kind));
  });
});

describe('renderChat', () => {
  it('streams the answer to stdout, status to stderr, exit 0 on completed', async () => {
    const { io, out, err } = captureIo();
    const code = await renderChat(readJsonRpcSse(streamFromString(FIXTURE, 64)), { json: false }, io);

    expect(code).toBe(0);
    expect(out()).toContain('CONSUMER_LAG');
    expect(out().endsWith('\n')).toBe(true);
    // status transitions land on stderr, not stdout
    expect(err()).toContain('working');
    expect(err()).toContain('completed');
    expect(out()).not.toContain('working');
  });

  it('--json emits raw events to stdout and no decoration', async () => {
    const { io, out, err } = captureIo();
    const code = await renderChat(readJsonRpcSse(streamFromString(FIXTURE, 4096)), { json: true }, io);

    expect(code).toBe(0);
    expect(err()).toBe('');
    const lines = out().trim().split('\n');
    expect(JSON.parse(lines[0]!).kind).toBe('task');
    expect(lines.every((l) => JSON.parse(l).kind !== undefined)).toBe(true);
  });

  it('exits non-zero and prints the error on a failed task', async () => {
    async function* failed(): AsyncGenerator<StreamEvent> {
      yield { kind: 'status-update', taskId: 't', contextId: 'c', status: { state: 'working', timestamp: '' }, final: false };
      yield {
        kind: 'status-update',
        taskId: 't',
        contextId: 'c',
        status: {
          state: 'failed',
          timestamp: '',
          message: { kind: 'message', role: 'agent', messageId: 'm', parts: [{ kind: 'text', text: 'Agent run failed: boom' }] },
        },
        final: true,
      };
    }
    const { io, err } = captureIo();
    const code = await renderChat(failed(), { json: false }, io);
    expect(code).toBe(1);
    expect(err()).toContain('failed');
    expect(err()).toContain('boom');
  });
});

describe('renderCard / renderTask', () => {
  const card: AgentCard = {
    protocolVersion: '1.0',
    name: 'Patch',
    description: 'desc',
    url: 'http://x/a2a',
    preferredTransport: 'JSONRPC',
    version: '0.1.0',
    provider: { organization: 'Datum', url: 'http://datum' },
    capabilities: { streaming: true, pushNotifications: false, stateTransitionHistory: false },
    defaultInputModes: ['text/plain'],
    defaultOutputModes: ['text/plain'],
    securitySchemes: { bearer: { type: 'http', scheme: 'bearer' } },
    security: [{ bearer: [] }],
    skills: [{ id: 'project-assistant', name: 'PA', description: 'd', tags: [] }],
  };

  it('pretty-prints the card and its json form', () => {
    const pretty = captureIo();
    renderCard(card, false, pretty.io);
    expect(pretty.out()).toContain('Patch');
    expect(pretty.out()).toContain('http://x/a2a');
    expect(pretty.out()).toContain('project-assistant');
    expect(pretty.out()).toContain('http bearer');

    const asJson = captureIo();
    renderCard(card, true, asJson.io);
    expect(JSON.parse(asJson.out()).name).toBe('Patch');
  });

  it('renders a task with its answer', () => {
    const { io, out } = captureIo();
    renderTask(
      {
        kind: 'task',
        id: 't-1',
        contextId: 'c-1',
        status: { state: 'completed', timestamp: '' },
        artifacts: [{ artifactId: 'response', parts: [{ kind: 'text', text: 'the answer' }] }],
      },
      false,
      io
    );
    expect(out()).toContain('t-1');
    expect(out()).toContain('completed');
    expect(out()).toContain('the answer');
  });
});

File: cmd/patch/render_test.go
package main

import (
	"encoding/json"
	"errors"
	"iter"
	"strings"
	"testing"

	"github.com/a2aproject/a2a-go/v2/a2a"
)

// capture is a test Io that records the two streams.
type capture struct {
	out strings.Builder
	err strings.Builder
}

func (c *capture) Out(text string) { c.out.WriteString(text) }
func (c *capture) Err(text string) { c.err.WriteString(text) }

// seqOf builds a streaming iterator from a fixed list of events.
func seqOf(events ...a2a.Event) iter.Seq2[a2a.Event, error] {
	return func(yield func(a2a.Event, error) bool) {
		for _, e := range events {
			if !yield(e, nil) {
				return
			}
		}
	}
}

func textPart(s string) *a2a.Part { return a2a.NewTextPart(s) }

func TestRenderChat_Completed(t *testing.T) {
	events := seqOf(
		&a2a.Task{ID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateSubmitted}},
		&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateWorking}},
		&a2a.TaskArtifactUpdateEvent{TaskID: "t", ContextID: "c", Artifact: &a2a.Artifact{ID: "response", Parts: a2a.ContentParts{textPart("Findings: CONSUMER_LAG")}}},
		&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateCompleted}},
	)
	var io capture
	code, err := renderChat(events, false, &io)
	if err != nil {
		t.Fatalf("unexpected stream err: %v", err)
	}
	if code != 0 {
		t.Errorf("code = %d, want 0", code)
	}
	// The answer text goes to stdout, ending with a tidy newline.
	if !strings.Contains(io.out.String(), "CONSUMER_LAG") {
		t.Errorf("stdout missing answer: %q", io.out.String())
	}
	if !strings.HasSuffix(io.out.String(), "\n") {
		t.Errorf("stdout should end with newline: %q", io.out.String())
	}
	// Status transitions go to stderr, not stdout.
	if !strings.Contains(io.err.String(), "working") || !strings.Contains(io.err.String(), "completed") {
		t.Errorf("stderr missing transitions: %q", io.err.String())
	}
	if strings.Contains(io.out.String(), "working") {
		t.Errorf("status leaked into stdout: %q", io.out.String())
	}
}

func TestRenderChat_JSON(t *testing.T) {
	events := seqOf(
		&a2a.Task{ID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateSubmitted}},
		&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateCompleted}},
	)
	var io capture
	code, err := renderChat(events, true, &io)
	if err != nil {
		t.Fatalf("unexpected stream err: %v", err)
	}
	if code != 0 {
		t.Errorf("code = %d, want 0", code)
	}
	// --json emits raw events to stdout, one per line, and no stderr decoration.
	if io.err.String() != "" {
		t.Errorf("stderr should be empty in json mode: %q", io.err.String())
	}
	lines := strings.Split(strings.TrimSpace(io.out.String()), "\n")
	if len(lines) != 2 {
		t.Fatalf("want 2 json lines, got %d: %q", len(lines), io.out.String())
	}
	for _, l := range lines {
		var v map[string]any
		if err := json.Unmarshal([]byte(l), &v); err != nil {
			t.Errorf("line is not valid JSON: %q (%v)", l, err)
		}
	}
}

func TestRenderChat_Failed(t *testing.T) {
	failMsg := a2a.NewMessage(a2a.MessageRoleAgent, textPart("Agent run failed: boom"))
	events := seqOf(
		&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateWorking}},
		&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateFailed, Message: failMsg}},
	)
	var io capture
	code, err := renderChat(events, false, &io)
	if err != nil {
		t.Fatalf("unexpected stream err: %v", err)
	}
	if code != 1 {
		t.Errorf("code = %d, want 1", code)
	}
	if !strings.Contains(io.err.String(), "failed") || !strings.Contains(io.err.String(), "boom") {
		t.Errorf("stderr missing failure detail: %q", io.err.String())
	}
}

func TestRenderChat_StreamError(t *testing.T) {
	boom := errors.New("stream exploded")
	events := func(yield func(a2a.Event, error) bool) {
		yield(&a2a.TaskStatusUpdateEvent{TaskID: "t", ContextID: "c", Status: a2a.TaskStatus{State: a2a.TaskStateWorking}}, nil)
		yield(nil, boom)
	}
	var io capture
	code, err := renderChat(events, false, &io)
	if code != 1 {
		t.Errorf("code = %d, want 1", code)
	}
	if !errors.Is(err, boom) {
		t.Errorf("want stream error surfaced, got %v", err)
	}
}

func TestRenderCard(t *testing.T) {
	card := &a2a.AgentCard{
		Name:        "Patch",
		Description: "desc",
		Version:     "0.1.0",
		SupportedInterfaces: []*a2a.AgentInterface{
			a2a.NewAgentInterface("http://x/a2a", a2a.TransportProtocolJSONRPC),
		},
		Provider:     &a2a.AgentProvider{Org: "Datum", URL: "http://datum"},
		Capabilities: a2a.AgentCapabilities{Streaming: true},
		SecuritySchemes: a2a.NamedSecuritySchemes{
			"bearer": a2a.HTTPAuthSecurityScheme{Scheme: "Bearer"},
		},
		Skills: []a2a.AgentSkill{{ID: "project-assistant", Name: "PA", Description: "d"}},
	}

	var pretty capture
	renderCard(card, false, &pretty)
	for _, want := range []string{"Patch", "http://x/a2a", "JSONRPC", "project-assistant", "http bearer", "Datum"} {
		if !strings.Contains(pretty.out.String(), want) {
			t.Errorf("pretty card missing %q:\n%s", want, pretty.out.String())
		}
	}

	var asJSON capture
	renderCard(card, true, &asJSON)
	var v map[string]any
	if err := json.Unmarshal([]byte(asJSON.out.String()), &v); err != nil {
		t.Fatalf("card json invalid: %v", err)
	}
	if v["name"] != "Patch" {
		t.Errorf("card json name = %v, want Patch", v["name"])
	}
}

func TestRenderTask(t *testing.T) {
	task := &a2a.Task{
		ID:        "t-1",
		ContextID: "c-1",
		Status:    a2a.TaskStatus{State: a2a.TaskStateCompleted},
		Artifacts: []*a2a.Artifact{{ID: "response", Parts: a2a.ContentParts{textPart("the answer")}}},
	}
	var io capture
	renderTask(task, false, &io)
	for _, want := range []string{"t-1", "completed", "the answer"} {
		if !strings.Contains(io.out.String(), want) {
			t.Errorf("task render missing %q:\n%s", want, io.out.String())
		}
	}
}
